oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/Traverser.java

```diff
@@ -169,5 +167,107 @@ public T next() {
             return current;
         }
     }
+
+    /**
+     * Returns an iterator that traverses a tree structure in post-order. Null nodes are strictly forbidden.
+     * <p>
+     * In post-order traversal, all children of a node are visited from left to right, followed by
+     * the node itself. This creates a bottom-up traversal pattern where leaf nodes are visited first,
+     * then their parents, and finally the root.
+     *
+     * @param <T> the type of value in the tree nodes
+     * @param root the root node of the tree, may be null
+     * @param childExtractor function to extract children from a node, must not be null
+     * @return an iterator that traverses the tree in post-order
+     * @throws NullPointerException if childExtractor or any child is null
+     */
+    public static <T> FluentIterable<T> postOrderTraversal(final T root, final @NotNull Function<T, Iterable<? extends T>> childExtractor) {
+        Objects.requireNonNull(childExtractor, "Children extractor function must not be null");
+        if (root == null) {
+            return FluentIterable.empty();
+        }
```

Comment:
Shouldn't it throw if root==null?

Reply:
@thomasmueller randomized test case has been added for this traversal order as well.

Reply:
@mbaedke Yes, you are right, Guava code does throw NPE in case the `root` is null. I will make the change to throw NPE in case the `root` is null.

Reply:
@mbaedke addressed in https://github.com/apache/jackrabbit-oak/pull/2447/commits/7e017e82d0d7cf433c6fe21e322e439d3eedb552